### DISCUSSION TITLE
Bake developer experience into Directus

### DISCUSSION THREAD

**Discussion by khanetor** [directus/directus#17308] [Feature Requests]
Directus is great and powerful, but I think there are a few rough corners with the developer experience. For example:
- Data migration seems manual, because, unlike most migration libraries, where we write or generate migration scripts at each stage, Directus generate a new independent schema snapshot in yaml, and Directus will somehow manage to update the schema from the new snapshot. This leads to un-unified, or non-standardized approaches to migration from dev to prod.
- Some important metadata, such as custom roles, or any change to theme, app name, and logo are not synced between environments.
- Flows, which is a hot topic, does not have a clear way to go from dev to prod environment.

All of these problems came from the fact that, in CD/CI, we share the code, and separate the data between environments. In Directus, however, we develop on data.

I think that, to make Directus an ultimate headless CMS platform, we should consider the use case of a dev environment, i.e. dev environment should be a (major) feature of Directus.

For example, when we setup Directus, we should specify 2 different database connections, 1 for dev, 1 for prod. From Directus perspective, these can be 2 equivalent environments, or one can be dedicated to dev, and the other for prod; we can decide on the detail in further discussion. Then, as a developer create a new collection, a new role, a new flow, there should be a button to "publish" the new collection, the new role, or the new flow to the other environment. The sync can be at schema level (to create new collection), or content level (to create new flow, new role, ...).

In this setup, dev and prod run on the same Directus instance, which is okay for small projects. For more ambitious projects, where the separation of dev and prod instances is a must, we can somehow make these 2 instances aware of each other, and the sync can happen in a similar fashion as mentioned above.

With this, Directus can become truly no-code, because there would not be any need for GitHub, or CI/CD. Hosted Directus can be much more appealing as well.¨

I hope this idea can contribute to Directus, and if I can join in this effort, please let me know.

**Comment by rijkvanzanten** [comment]
Love this idea @nlhkh! An important first step to support this is to allow multiple databases to be connected at the same time, after which we can implement the logic to allow copying / migrating parts from one to the other 🙂

**Reply by khanetor** [reply]
Is the context for multiple database connections for switching between different environments?

**Reply by rijkvanzanten** [reply]
It's not the only use case, but would definitely make something like this possible 🙂

**Comment by richardturnerfos** [comment]
@nlhkh I couldn't agree more. Directus is an amazing tool, and we're still trying to work out how it can be part of our process and how we can support it. Just found snapshots and got tripped up by the flows not coming across. (In my simplistic mind, it's like the tables of a database being migrated by not the stored procedures). Currently "Go Live" would mean sitting a developer in front of a live server and hoping they don't forget every little change added to Dev. 

I'm a little concerned about having one site communicate with multiple databases for migrations. Including an "exportable migration script" option or expanding snapshots would be preferable, reviewable, and trackable. There are plenty of places where prod is firewalled off, managed by separate teams, or deployed in other companies' environments.  Here, Dev and Prod sites wouldn't be allowed to communicate. The Multiple Database idea would be great for Dev->Dev and Dev->Staging/Beta environments.

**Comment by falmanna** [comment]
I agree with all the mentioned limitations, but I have different perspective on the solutions. 

I don't think directus should care about dev, prod, and environments in general, what should be improved are the CLI/API utils. Recently we got many improvements on the schema migration front, which is a great step towards a proper DEX and CI/CD integration. 

Few of the things that I think can improve and lead toward a great CI/CD experience:
- Incremental migrations script generation: Currently, most users change data directly in the data studio, few write custom migrations. To solve for both, we need a way to generate incremental migrations that can take the DB schema to a pre-defined state. Similar to Django and Prisma. 
- Incremental system data migration: Flows and Roles and other system configurations resides in DB and requires manual sync between envs. What can be done is a utility to incrementally export/import system data. 
- All these utils should be accessible from both the CLI and the API for cases when directus is running on serverless env without CLI access.

Having all these, the developer experience would be: 
1- Work on a local directus instance, generate schema/data migrations, push to PR. 
2- Tests and checks dry run the generated migrations to detect conflicts
3- Team reviews and merge on approve.  
4- CI/CD kicks in and applied the changes
5- prod deploy applies the full check, test, migrate flow.

This also can improve replacing DB providers transparently or in moving to the cloud directus offering.

**Reply by claytongulick** [reply]
Are you suggesting adding the ability to the snapshot functionality of doing just a partial snapshot? Maybe limited to specific collections/fields? Then those partial snapshots could be executed directly, or stored in a migrations folder and run in-order?

If so, I think maybe in addition to the cli, the schema settings UI could be enhanced to show "dirty" indicators for schema changes that have been made since the last snapshot was synced. And have selections to generate the partial snapshot yaml files.

I like it.

**Reply by falmanna** [reply]
@claytongulick Yeah, more or less. 
So whenever you run the snapshot/migration util, you get a new timestamped migration file (similar to the ones in extensions). There, you can find the changes, and also manage data migrations with the `up` and `down` functions. This is a stretch and I know the current snapshot tool is limited to schema changes, but it is an important part.

Not sure about the UI as directus might be used by non-developers for data management work, so this might confuse them.

I created this helper script to give me some of these functionalities, and I plan to migrate it to use the schema APIs once released. 
```
#!/bin/bash

# generate schema function
function generate_schema() {
    npx directus schema snapshot ./schema/latest.yaml --format yaml -y
}

# apply schema function
function apply_schema() {
    npx directus schema apply ./schema/latest.yaml -y
}

# dry run schema function
function dry_run_schema() {
    npx directus schema apply ./schema/latest.yaml -y --dry-run
}

# check schema function, it will generate a new schema and compare it with the latest one
function check_schema() {
    npx directus schema snapshot ./schema/temp.yaml --format yaml -y
    # find if files are different and delete the temp file. Raise error if files are different
    if ! cmp -s "./schema/latest.yaml" "./schema/temp.yaml"; then
        echo "Schema is not up to date"
        rm ./schema/temp.yaml
        exit 1
    else
        echo "Schema is up to date"
        rm ./schema/temp.yaml
        exit 0
    fi
}

# accept generate, check, apply, dry-run parameters as input to this script
# if no parameters are passed, then return help

if [ $# -eq 0 ]
then
    echo "No arguments supplied"
    echo "Usage: schema-utils.sh [generate|check|apply|dry-run]"
    exit 1
fi

# call the function based on the input parameter
case $1 in
    generate)
        generate_schema
    ;;
    check)
        check_schema
    ;;
    apply)
        apply_schema
    ;;
    dry-run)
        dry_run_schema
    ;;
    *)
        echo "Invalid argument"
        echo "Usage: ./schema-utils.sh [generate|check|apply|dry-run]"
        exit 1
    ;;
esac

```

**Reply by rijkvanzanten** [reply]
> Are you suggesting adding the ability to the snapshot functionality of doing just a partial snapshot? Maybe limited to specific collections/fields? Then those partial snapshots could be executed directly, or stored in a migrations folder and run in-order?

FWIW the new schema endpoints (https://github.com/directus/directus/pull/14713)—which will be released very soon—support partial migrations as well 🙂

**Reply by luochuanyuewu** [reply]
Link broken> > Are you suggesting adding the ability to the snapshot functionality of doing just a partial snapshot? Maybe limited to specific collections/fields? Then those partial snapshots could be executed directly, or stored in a migrations folder and run in-order?
> 
> 
> 
> FWIW the new schema endpoints (https://github.com/directus/directus/pull/14713)—which will be released very soon—support partial migrations as well 🙂

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
